vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -162,21 +211,255 @@ void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+        for (int c = 1; c <= static_cast<int>(cols); c++) {
+            p_placement.node_loc_x[(r-1)*cols + c -1] = c * x_gap;
+            p_placement.node_loc_x[(r-1)*cols + c -1] = r * y_gap;
         }
     }
-    return true;
+}
+
+// This function return the two nodes on the bound of a netlist, (max, min)
+std::pair<size_t, size_t> B2BSolver::boundNode(std::vector<size_t>& node_ids, std::vector<double>& node_locs){
+    auto compare = [&node_locs](int a, int b) {
+        return node_locs[a] < node_locs[b];
+    };
+    auto maxIt = std::max_element(node_ids.begin(), node_ids.end(), compare);
+    auto minIt = std::min_element(node_ids.begin(), node_ids.end(), compare);
+    VTR_ASSERT(node_locs[*maxIt] >= node_locs[*minIt] && "max note should be greater than min node");
+    return std::make_pair(*maxIt, *minIt);
+}
+
+void B2BSolver::populate_matrix(PartialPlacement &p_placement) {
+    const AtomContext& atom_ctx = g_vpr_ctx.atom();
+    const AtomNetlist& netlist = p_placement.atom_netlist;
+    // Resetting As bs
+    A_sparse_x.setZero();
+    A_sparse_y.setZero();
+    std::vector<Eigen::Triplet<double>> tripletList_x;
+    tripletList_x.reserve(p_placement.num_moveable_nodes * netlist.nets().size());
+    std::vector<Eigen::Triplet<double>> tripletList_y;
+    tripletList_y.reserve(p_placement.num_moveable_nodes * netlist.nets().size());
+    b_x = Eigen::VectorXd::Zero(p_placement.num_moveable_nodes);
+    b_y = Eigen::VectorXd::Zero(p_placement.num_moveable_nodes);
+
+    for (AtomNetId net_id : netlist.nets()) {
+        if (p_placement.net_is_ignored_for_placement(net_id))
+            continue;
+
+        int num_pins = netlist.net_pins(net_id).size();
+        VTR_ASSERT(num_pins > 1 && "net least has at least 2 pins");
+        
+        std::vector<size_t> node_ids;
+        for (AtomPinId pin_id : netlist.net_pins(net_id)) {
+            AtomBlockId blk_id = netlist.pin_block(pin_id);
+            size_t node_id = p_placement.get_node_id_from_blk(blk_id, atom_ctx.atom_molecules);
+            node_ids.push_back(node_id);
+        }
+        // remove duplicated node, they are there becaues of prepacked molecules.
+        std::set<size_t> node_ids_set(node_ids.begin(), node_ids.end());
+        std::vector<size_t> node_ids_no_duplicate(node_ids_set.begin(), node_ids_set.end());
+        
+        auto [maxXId, minXId] = boundNode(node_ids_no_duplicate, p_placement.node_loc_x);
+        auto [maxYId, minYId] = boundNode(node_ids_no_duplicate, p_placement.node_loc_y);
+        if (node_ids_no_duplicate.size() <= 1){
+            continue;
+        }
+        // assign arbitrary node as bound node when they are all equal
+        if (maxXId == minXId) {
+            maxXId = node_ids_no_duplicate[0];
+            minXId = node_ids_no_duplicate[1];
+        }
+        if (maxYId == minYId) {
+            maxYId = node_ids_no_duplicate[0];
+            minYId = node_ids_no_duplicate[1];
```

Comment:
I am nervous about how you are breaking these ties. Although its deterministic technically it can cause problems since the order is determined by a set being converted to a vector. I would put a TODO comment to investigate better handling ties.